stellargraph/layer/knowledge_graph.py

```diff
@@ -137,6 +137,96 @@ def embeddings(model):
+        score 3.14, and only one modified-object edge has a higher score (e.g. ``("a", "X", "c")``),
+        then the raw modified-object rank for ``E`` will be 2; if all of the ``(n, "X", "b")`` edges
+        have score less than 3.14, then the raw modified-subject rank for ``E`` will be 1.
+
+        Args:
+            model (tensorflow.keras.Model): a Keras model created using a ``ComplEx`` instance.
+
+            test_data: the output of :meth:`KGTripleGenerator.flow` on some test triples
+
+            known_edges_graph (StellarGraph):
+                a graph instance containing all known edges/triples
+
+        Returns:
+            A numpy array of integer raw ranks. It has shape ``N × 2``, where N is the number of
+            test triples in ``test_data``; the first column (``array[:, 0]``) holds the
+            modified-object ranks, and the second (``array[:, 1]``) holds the modified-subject
+            ranks.
+        """
+
+        if not isinstance(test_data, KGTripleSequence):
+            raise TypeError(
+                "test_data: expected KGTripleSequence; found {type(test_data).__name__}"
+            )
+
+        num_nodes = known_edges_graph.number_of_nodes()
+
+        def ranks(pred, true_ilocs, true_is_source):
+            batch_size = len(true_ilocs)
+            assert pred.shape == (num_nodes, batch_size)
+
+            # the score of the true edge, for each edge in the batch (this indexes in lock-step,
+            # i.e. [pred[true_ilocs[0], range(batch_size)[0]], ...])
+            true_scores = pred[true_ilocs, range(batch_size)]
+
+            # for each column, compare all the scores against the score of the true edge
+            greater = pred > true_scores
+
+            # the raw rank is the number of elements scored higher than the true edge
+            raw_rank = 1 + greater.sum(axis=0)
+            assert raw_rank.shape == (batch_size,)
+
+            return raw_rank
+
+        n, all_r = ComplEx.embeddings(model)
+        n_conj = n.conj()
+
+        raws = []
+
+        # run through the batches and compute the ranks for each one
+        num_tested = 0
+        for i in range(len(test_data)):
+            ((subjects, rels, objects),) = test_data[i]
+            num_tested += len(subjects)
+
+            # batch_size x k
+            ss = n[subjects, :]
+            rs = all_r[rels, :]
+            os = n[objects, :]
+
+            # reproduce the scoring function for ranking the given subject and relation against all
+            # other nodes (objects), and similarly given relation and object against all
+            # subjects. The bulk operations give speeeeeeeeed.
+            # (num_nodes x k, batch_size x k) -> num_nodes x batch_size
+            mod_o_pred = np.inner(n_conj, ss * rs).real
+            mod_s_pred = np.inner(n, rs * os.conj()).real
+
+            mod_o_raw = ranks(mod_o_pred, true_ilocs=objects, true_is_source=True)
```

Comment:
`true_is_source` seems a bit confusing to me. when `true_is_source=True` the `true_ilocs` are the `objects`, when it seems to me that it should be the other way around. Can you explain why you chose this?

Reply:
It's only relevant for the filtered ranks, so I'm going to remove it from this PR... however, maybe this is why the filtering wasn't behaving as expected.